### PR TITLE
[Diagnostics] Improve keypath diagnostic  when cannot be inferred by context

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
Improving the diagnostics when the key path type can't be inferred by context. e.g. `let _ : AnyKeyPath = \.x`

cc @xedin @hborla @varungandhi-apple :)

### CODE REVIEW

File: include/swift/AST/DiagnosticsSema.def
Comment (xwu) [inline include/swift/AST/DiagnosticsSema.def:62]:
```suggestion
      "did you mean to use a key path subscript?",
```

Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def]: Thanks @xwu!

File: test/expr/unary/keypath/keypath.swift
Comment (varungandhi-apple) [inline test/expr/unary/keypath/keypath.swift:903]: Thanks for the PR! IIUC, there are three possibilities:

1. The `\.x` does not match any of the explicitly defined subscripts (possibly 0): In this case, the user most likely meant to write an explicit `keyPath: `.
2. The `\.x` matches multiple subscripts which means that the type is ambiguous from the context. It is still possible the the user meant to write an explicit `keyPath:` but less likely than in case 1.
3. There's something else going wrong which is hard to judge (e.g. the user meant to type `x` referring to a local variable instead of `\.x`).

My guess is that case 1 is probably the most common in practice. By adding an extra "cannot infer contextual type for key path; consider adding an explicit root type", that makes it seem like there are 2 different problems to be fixed (the "root type problem" and the "key path subscript problem"), even though just fixing the second one will most likely fix the issue. I think we should only issue the warning about the contextual type here if we are confident that adding the `keyPath:` will not fix the issue.

File: test/expr/unary/keypath/keypath.swift
Comment (varungandhi-apple) [inline test/expr/unary/keypath/keypath.swift:909]: Not sure why we are checking the diagnostic for `s[1]` here... We already have several tests for "has no subscripts" elsewhere.

File: include/swift/AST/DiagnosticsSema.def
Comment (varungandhi-apple) [inline include/swift/AST/DiagnosticsSema.def:80]: Thoughts on dropping "contextual" and "explicit"?

```
cannot infer type for key path; consider adding a root type
```

File: include/swift/AST/DiagnosticsSema.def
Comment (varungandhi-apple) [inline include/swift/AST/DiagnosticsSema.def:63]: Will this diagnostic be emitted even when the type might have explicit non-keypath subscripts? I can't tell by looking at the code directly; I would guess that it should be (or at least, the "did you mean to use a key path subscript" would still be applicable), and if that's the case, then the phrasing needs to avoid saying "has no subscripts". Not sure how we could phrase it better though...

File: test/expr/unary/keypath/keypath.swift
Comment (LucianoPAlmeida) [reply test/expr/unary/keypath/keypath.swift:909]: The idea was just to show the general case in the same block since its really simple, but I can remove it no problem.

**Comment by varungandhi-apple** [comment]
@swift-ci please test

File: test/expr/unary/keypath/keypath.swift
Comment (LucianoPAlmeida) [reply test/expr/unary/keypath/keypath.swift:903]: Humm, the thought here was to show all that's going on in the context, because indeed the solver could not infer the root from the context in this case. I think is fine to show it as an error because it is indeed not applicable

File: include/swift/AST/DiagnosticsSema.def
Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def:80]: I really not sure, I think is fine to change it, but I think the actual verbosity makes a bit more clear.
But can change no problem :)

File: include/swift/AST/DiagnosticsSema.def
Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def:63]: That's a good question, let me see if I can improve this situation :)

File: test/expr/unary/keypath/keypath.swift
Comment (LucianoPAlmeida) [reply test/expr/unary/keypath/keypath.swift]: Fixed!

File: include/swift/AST/DiagnosticsSema.def
Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def]: I think the new message makes more sense, related to cases where we have 1+ subscripts on the base type, there is something else going on I can't still figure it out. Maybe let it for another patch ... let's try to figure out these ones first :)

Comment (varungandhi-apple) [reply include/swift/AST/DiagnosticsSema.def]: Thanks, the new diagnostic avoids the potentially incorrect verbiage.

File: include/swift/AST/DiagnosticsSema.def
Comment (varungandhi-apple) [reply include/swift/AST/DiagnosticsSema.def:80]: That's a good reason to have a discussion. I'm not trying to force people into following what I say. Here's my thought process. Please feel free to disagree and expand on why you think those terms make the diagnostic clearer. That's how we make things better. 🙂 

With "explicit root types", there's the (probably small) additional risk of an wrong turn where someone might think "what is the difference from a root type and an explicit root type" (even though what we actually meant was "adding a root type explicitly"). (FWIW, I Google-searched for "Swift key path root type" and "Swift key path explicit root type" and the former gave better search results.)

With "contextual type" it's likely that the programmer has seen an error mentioning contextual type before and the way they typically fix it is by actually adding a type annotation for a binding (or in some cases using `as`)

```
let x = .mycase -> let x : MyType = .mycase
```

So they might think "how should I add a type annotation with `:`" or they might think "should I add a ", if they hung up on the first part of the diagnostic (maybe because they don't fully understand what the second part of the diagnostic is trying to say). But both of those solutions are suboptimal.

Again, this is probably a small risk, but it's a potential thing that can go wrong.

What we really want is the person to understand that (a) there is a type inference failure (as opposed to many other kinds of failures) (b) adding a root type to the key path expressions will fix the problem. I don't think the terms "contextual" or "explicit" are justifying the potential cognitive cost, because they're not advancing any of these sub-goals IMO. The verbiage without them

```
cannot infer type for key path; consider adding a root type
```

is better IMO in that both sub-goals are tackled with less room for following an incorrect line of reasoning.

File: include/swift/AST/DiagnosticsSema.def
Comment (varungandhi-apple) [reply include/swift/AST/DiagnosticsSema.def]: > related to cases where we have 1+ subscripts on the base type, there is something else going on I can't still figure it out. Maybe let it for another patch 

Could you leave a `TODO` (either near the diagnostic or where it is being emitted right now) about this, so that it is not forgotten?

File: test/expr/unary/keypath/keypath.swift
Comment (varungandhi-apple) [reply test/expr/unary/keypath/keypath.swift]: Ok, now I see where you're coming from. Here's my perspective: I think when a programmer sees an error, they most likely do not want to see all the information available. They want to understand what the error is, why there is an error (i.e. understand the contextual information that is contributing to the error), and hopefully some accurate hints on how to fix it.

In this particular case there's no "one true error", there's multiple possible errors, which is what is making things difficult. If there are multiple possible errors, it should be made clear that we are talking about `error1 || error2`, not `error1 && error2`. IMO showing all of them, unless absolutely necessary, is more likely to lead to frustration (because it makes the programmer think `&&` instead of `||`). Worse, when they see two seemingly unrelated errors at the same spot, and fixing one seemingly fixes both of them, that erodes trust in the compiler. The more often it happens, the more likely the programmer is to start ignoring the diagnostics.

In this particular case, I think there's three main possibilities which I pointed out earlier (maybe there's more which I missed). We can't really address the third one.

Since there is a likelihood that one of the possible errors (first case) is much more likely than the other, I think it's valuable to optimize the diagnostic for that case. This means pruning out inapplicable information from the context (the error for the second case); more information is more room for distraction; if we know that the information is not going to be applicable, we should avoid emitting it.

That said, I don't know how hard implementing the check (to only omit one error instead of both, when applicable) is. I will defer to you and others on that.

File: include/swift/AST/DiagnosticsSema.def
Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def]: Humm that makes sense, thank you for the detailed description :) 
I just changed, adding an extra `from context` to make it clear that it is failing to infer from the current context the expression is in. `cannot infer type for key path;` may give room to think that the compiler cannot infer key path types at all(although is obviously not true). 
The message is `cannot infer key path type from context; consider adding a root type` let me know what you think :)

Comment (varungandhi-apple) [reply include/swift/AST/DiagnosticsSema.def]: Thanks. I see your point about potentially making someone think that the compiler can't infer those types generally, that's fair. I think if someone's using a key path, it is likely that they've probably already used it in contexts where the type was inferred for them, so maybe it's not such a big risk.

File: test/expr/unary/keypath/keypath.swift
Comment (LucianoPAlmeida) [inline test/expr/unary/keypath/keypath.swift:918]: Not sure if we should diagnose anything else here since `repairFailures` is considering an `ApplyArgToParam` failure involving a hole as fixed ...
https://github.com/apple/swift/blob/7d2f9440866c5360cc4929f003850058ecfc98bc/lib/Sema/CSSimplify.cpp#L3619
cc @xedin

File: include/swift/AST/DiagnosticsSema.def
Comment (xwu) [reply include/swift/AST/DiagnosticsSema.def]: If we want to be precise, one isn't "adding" a root type (there always has to be one), only specifying that type explicitly. To avoid confusion you could write "consider explicitly adding a root type."

Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def]: I like it that way, feels precise :)

**Review by xedin — COMMENTED** [review]
Thanks for looking into this, @LucianoPAlmeida!

Couple of thoughts:

- I think we should split this into two separate PRs, one for contextual stuff and other one for `keyPath:` overload.
- Regarding `keyPath:`
   - I think we can do better in this case and instead of saying `maybe` we could actually verify that `keyPath:` overload choice works (good that it doesn't have to be a declaration :)) and suggest adding `keyPath:` label.

- Regarding root/value inference failure:
  - Is it any different from a regular inability to infer generic arguments? 
  - I think we should add a special diagnostic about `AnyKeyPath` which state that it doesn't
    provide enough contextual information to allow root type to be inferred implicitly 
    (because its root type is `Any`). That should help resolve at least some confusion here...
  - Could we suggest a fix-it in this case to specify `<#Root#>` before a leading dot ?

**Comment by LucianoPAlmeida** [comment]
> * I think we should split this into two separate PRs, one for contextual stuff and other one for `keyPath:` overload.

Sure! Let's handle contextual first here then :)

>   * I think we can do better in this case and instead of saying `maybe` we could actually verify that `keyPath:` overload choice works (good that it doesn't have to be a declaration :)) and 
suggest adding `keyPath:` label.

 Noted :)

>   * Is it any different from a regular inability to infer generic arguments?

I didn't thought about that, perhaps there is no difference... 
but I think we can tailor the diagnostics for key paths since generic arguments is more general...? 

>   * I think we should add a special diagnostic about `AnyKeyPath` which state that it doesn't
>     provide enough contextual information to allow root type to be inferred implicitly
>     (because its root type is `Any`). That should help resolve at least some confusion here...

>   * Could we suggest a fix-it, in this case, to specify `<#Root#>` before a leading dot ?

I'll work on those, thanks @xedin :)

**Comment by xedin** [comment]
> I think we can tailor the diagnostics for key paths since generic arguments is more general.

I agree that we can tailor diagnostic messages but I think emission of the fix/diagnostic should be based on what we do for generic parameters.

**Comment by artemcm** [comment]
I was accidentally concurrently looking into this issue. From what I gather, this PR is now resolving the more general issue with key path types not being inferred from context, but not the issue of the `keyPath:` label overload?

**Comment by LucianoPAlmeida** [comment]
> I was accidentally concurrently looking into this issue. From what I gather, this PR is now resolving the more general issue with key path types not being inferred from context, but not the issue of the `keyPath:` label overload?

Initially, it was trying to handle both, but I just remove the commit since @xedin suggested split in two PRs, will do a follow up as soon as this one lands :)

**Comment by artemcm** [comment]
@LucianoPAlmeida, I can't see the removed commits, but please take a peak at #31716, I imagine it is something similar?

**Comment by varungandhi-apple** [comment]
I'm a bit confused as to what's going to happen here? Are we merging Artem's PR or this one? I can't comment on the implementation but I think the diagnostic looks good. 👍 

@swift-ci please test

**Comment by LucianoPAlmeida** [comment]
> I'm a bit confused as to what's going to happen here? Are we merging Artem's PR or this one? I can't comment on the implementation but I think the diagnostic looks good. 👍

@varungandhi-apple We decided to split this PR into two parts contextual inference issues, and subscript fixes as a follow-up. So we let this one only to solve contextual issues and planning a follow up for the keyPath label... but coincidentally the second part is already handled by Artem's PR :)

**Comment by LucianoPAlmeida** [comment]
I think Artem's covers PR the whole scope of the SR, so I removed the test cases related and will let this be only contextual stuff :)

File: lib/Sema/CSFix.h
Comment (xedin) [inline lib/Sema/CSFix.h:264]: `by` -> `from`.

File: include/swift/AST/DiagnosticsSema.def
Comment (xedin) [inline include/swift/AST/DiagnosticsSema.def:78]: @hborla What is better - `consider explicitly adding ...` or `consider adding ... explicitly`?

**Comment by xedin** [comment]
@LucianoPAlmeida I think it's `from` context instead of `by` context but I might be wrong...

File: lib/Sema/CSFix.h
Comment (LucianoPAlmeida) [reply lib/Sema/CSFix.h]: Right! Fixed :)

File: include/swift/AST/DiagnosticsSema.def
Comment (hborla) [reply include/swift/AST/DiagnosticsSema.def:78]: What about `consider explicitly specifying a root type`? That's consistent with the language in other error messages prompting the user to add an explicit type somewhere

I haven't caught up on the other conversation in this PR, but is there a reason we aren't providing this as a fix-it to insert a placeholder for the root type?

Comment (xedin) [reply include/swift/AST/DiagnosticsSema.def:78]: I like that wording better, thank you! And to answer your question - we do provide a fix-it to add a placeholder for root type.

File: lib/Sema/CSFix.h
Comment (xedin) [inline lib/Sema/CSFix.h:1830]: Nit: should probably be ‘final`

File: include/swift/AST/DiagnosticsSema.def
Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def:78]: Right! I agree this feels better. Thanks @hborla :)

Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def]: Fixed :)

**Review by xedin — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test

**Pull request merged**
